### DISCUSSION TITLE
Compiling apps together

### DISCUSSION THREAD

**Discussion by rtaylo45** [idaholab/moose#18684] [Q&A Getting Started] [answered]
Hello,

I have a problem were I am compiling two moose apps together and i am getting a linking error: 

```
Linking Library /home/qt5/projects/mole/test/lib/libmole_test-opt.la...
libtool:   error: cannot find the library '/home/qt5/projects/mole/test/lib/libmole_test-opt.la' or unhandled argument '/home/qt5/projects/mole/test/lib/libmole_test-opt.la'
```

This is what my make file looks like:
```
# Set up MOOSE framework
ifeq ($(wildcard $(MOOSE_DIR)/framework/Makefile),)
  $(error MOOSE framework was not found!  Please set MOOSE_DIR to the MOOSE directory!)
endif
FRAMEWORK_DIR      := $(MOOSE_DIR)/framework
include $(FRAMEWORK_DIR)/build.mk
include $(FRAMEWORK_DIR)/moose.mk

################################## MODULES ####################################
# To use certain physics included with MOOSE, set variables below to
# yes as needed.  Or set ALL_MODULES to yes to turn on everything (overrides
# other set variables).

ALL_MODULES                 := no

CHEMICAL_REACTIONS          := yes
CONTACT                     := no
EXTERNAL_PETSC_SOLVER       := no
FLUID_PROPERTIES            := no
FUNCTIONAL_EXPANSION_TOOLS  := no
HEAT_CONDUCTION             := no
LEVEL_SET                   := yes
MISC                        := no
NAVIER_STOKES               := no
PHASE_FIELD                 := yes
POROUS_FLOW                 := no
RDG                         := no
RICHARDS                    := no
SOLID_MECHANICS             := no
STOCHASTIC_TOOLS            := no
TENSOR_MECHANICS            := no
XFEM                        := no

# Modify modules for SAM
ifneq ($(wildcard $(SAM_DIR)/Makefile),)
  # Sam will need these modules
  FLUID_PROPERTIES := yes
  HEAT_CONDUCTION  := yes
  TENSOR_MECHANICS := yes
endif

include $(MOOSE_DIR)/modules/modules.mk

############################## SAM ############################################
ifneq ($(wildcard $(SAM_DIR)/Makefile),)
  APPLICATION_DIR     := ${SAM_DIR}
  APPLICATION_NAME    := sam
  BUILD_EXEC          := yes
  include             $(FRAMEWORK_DIR)/app.mk
  ADDITIONAL_INCLUDES += -I$(SAM_DIR)/include/base
  ADDITIONAL_CPPFLAGS += -DSAM_ENABLED
  ADDITIONAL_LIBS     += $(SAM_DIR)/lib/libsam-$(METHOD).la
endif

############################## MOLE ############################################
APPLICATION_DIR    := $(CURDIR)
APPLICATION_NAME   := mole
BUILD_EXEC         := yes
GEN_REVISION       := no
DEP_APPS           := $(shell $(FRAMEWORK_DIR)/scripts/find_dep_apps.py $(APPLICATION_NAME))
include            $(FRAMEWORK_DIR)/app.mk

###############################################################################
# Additional special case targets should be added here
```
I thought i got this working a few months ago but I can't remember if i did anything to the Makefile. Im compiling both codes with MOOSE commit 00fa7c0bbe.

**Comment by GiudGiud** [comment, marked as answer]
I think you may be missing
BUILD_TEST_OBJECTS_LIB := yes in your if_SAM loop
?

Guillaume

**Reply by rtaylo45** [reply]
So that did not help. Whats interesting is that if i try to recompile it right after it fails to compile the first time. It works. The first time i try to compile it, it says that it cannot find `libmole_test-opt.la` but if i look in the folder it was looking in, its there. So when i compile it again, it finds the library.

**Reply by GiudGiud** [reply]
Let's look at the output of make -n and see in what order things are being linked. You could email it to me if you dont want to share it here

**Reply by permcody** [reply]
Zack, A few things which may or may not help:
- You should never need to touch the `main.C` file. If you think you need to, you are working outside of our intended system. 
- Your Makefile seems a little bit complicated. We've really tried to simplify the necessary blocks you need when compiling applications together. If the dependent MOOSE application is required, the block should be this simple:
```
# SAM
SAM_DIR            ?= $(CURDIR)/SAM
APPLICATION_DIR    := $(SAM_DIR)
APPLICATION_NAME   := sam
include            $(FRAMEWORK_DIR)/app.mk
```

For optional dependencies, it does get a bit more complex since you have to create compiler defines and use several preprocessor commands as you've done. 

One other big difference is that we normally don't build "execs" for anything but the final application. While I can't see why that would be the problem it's related to the "test" library that you are getting so maybe. Perhaps a variable we expect to only be defined once is getting defined multiple times (again, our system expects only a single exec). 

The test library should be built only for the final application it's normally not linked in anywhere except for that final application. The intent is that you can have test objects that don't pollute scope further up the stream. Try taking that line out and see if we can get further.

We should chat about getting access at the INL. I think it would be very helpful. We can setup a government use agreement very quickly. It doesn't have to go through full technology deployment process for another lab to get it.

**Reply by rtaylo45** [reply]
I made the changes to the Makefile as you suggested and the code compiles as it should with no errors. I reverted back my changes made in the `main.C` file. So i changed from using the `moleApp` object to the original `moleTestApp` object. I have a set of test that run for SAM/Mole coupled which failed because it was not able to find the Actions/kernels.. etc. that should be provided by SAM. I also tried running the test files with `mole-opt` exc and i get the same failure. If i made the changes to the  `main.C` file to use the `moleApp` object instead of `moleTestApp` it works. I would like to note that we changes to the `moleApp.C` file  for special cases when SAM is compiled with Mole. These changes include:

```
InputParameters
moleApp::validParams()
{
  auto params = MooseApp::validParams();
#ifdef SAM_ENABLED
  params += SamApp::validParams();
#endif

  // Class description
  params.addClassDescription("The moleApp class creates an instance of a mole "
    "solver, which solves various species transport related problems for "
    "molten salt reactors.");

  return params;
}
```

```
moleApp::moleApp(InputParameters parameters)
#ifdef SAM_ENABLED
  : SamApp(parameters)
#else
  : MooseApp(parameters)
#endif
{
  moleApp::registerAll(_factory, _action_factory, _syntax);
}
```

```
void
moleApp::registerAll(Factory & f, ActionFactory & af, Syntax & s)
{
  ModulesApp::registerAll(f, af, s);
  Registry::registerObjectsTo(f, {"moleApp"});
  Registry::registerActionsTo(af, {"moleApp"});

  /* register custom execute flags, action syntax, etc. here */
  associateSyntaxInner(s, af);
#ifdef SAM_ENABLED
  SamApp::registerAll(f, af, s);
#endif
}
```

Would these changes also need to be included in the `moleTestApp.C` file? Why does the `main.C` file use the `moleTestApp` object and not the `moleApp` object?

**Reply by GiudGiud** [reply]
Some of the changes (like registering from SAMApp) will be included because moleTestApp::registerAll calls moleApp::registerAll. 
Same for validParams, moleTestApp gets its validParam from moleApp 

The only change I can see that doesnt propagate is moleApp inheriting from SamApp, since moleTestApp does not inherit from moleApp. 
What do you need that for?

**Reply by rtaylo45** [reply]
We thought it would be required to have the SAM app kerenels/actions available to Mole so that we could use them in the input file. I came into this project after the initial work was done. So i did not add this part

**Reply by rtaylo45** [reply]
I changed the moleTestApp object to inherit from SamApp like I did with molApp and everything seems to work. Thought, I don't really like having to inherit from the SamApp.

**Reply by GiudGiud** [reply]
you should not need that, but I just went to check and BlueCrab (the ultimate combined app) does that (inheriting from SamApp too). 
So it is what it is..
Ok glad to hear that. Please let us know if any more issues arise

**Reply by aarograh** [reply]
@GiudGiud (or any other MOOSE dev) I have been working with @rtaylo45 and others on development of our Mole app.  This issue has arisen again because we are attempting to couple with several different codes. and a couple different developers ended up pursuing this different ways.

I have a couple questions.  First, a "clean" moleApp will have `class moleApp : public MooseApp` and `class moleTestApp : public MooseApp`.  As Zack mentioned, we modified the first of those to be `class moleApp : public SamApp`.  What I don't understand is this: moleTestApp can properly use moleApp objects just fine, but it cannot use SamApp objects (unless we have it also inherit from SamApp).  However, moleTestApp does not inherit from moleApp.  I do not understand why it can successfully use all the mole objects but cannot use the Sam objects?  Especially since moleApp does all the objects and app registrations for Sam as well.  If you could provide some additional clarity on exactly what the expectations are, that woudl be helpful.  I want to prevent this kind of thing:
```c++
#ifdef APP1_ENABLED && APP2_ENABLED && APP3_ENABLED
  class moleApp : public app1, public app2, public app3
#elseif APP1_ENABLED && APP2_ENABLED
  class moleApp : public app1, public app2
#elseif APP1_ENABLED && APP3_ENABLED
  class moleApp : public app1, public app3
...
etc.
...
```
Obviously that gets silly after awhile.  And furthermore, having to make this same set of changes in both moleApp and moleTestApp is 1) annoying and redundant, and 2) confusing to me.

The second question I have concerns the test objects.  It's unclear to me exactly what kinds of objects this might be?  I'm envisioning special post-processors, problems, or executioners to operate the application in an abnormal way for testing, but I've never made use of such a setup.  Is there useful documentation/examples of what such test objects might be?  It's possible that we may not use them, in which case we may take the simple approach of modifying main.C to just use moleApp instead of moleTestApp.  However, I do not want to make this change without first understanding possible issues we may encounter as a results.

**Reply by GiudGiud** [reply]
Hello

A partial response to your questions:
- If you dont do this:
```
#ifdef APP1_ENABLED && APP2_ENABLED && APP3_ENABLED
  class moleApp : public app1, public app2, public app3
#elseif APP1_ENABLED && APP2_ENABLED
  class moleApp : public app1, public app2
#elseif APP1_ENABLED && APP3_ENABLED
  class moleApp : public app1, public app3
...
etc.
```
how is the compiler going to know what a SamApp is when Sam is not enabled? 
If you choose to work with inheritance (and one alternative is dynamic loading), then you have to do this imo. 
@andrsd may advise you better here, he is in charge of another combined application. 

Note that for the test app, it calls the registering routines from the main app, see there for example
https://github.com/idaholab/TMAP8/blob/devel/test/src/base/TMAPTestApp.C

I dont see why test apps could not inherit from the main app, but i have not tried this setup.

- Test objects are for reaching hard-to-reach statements in the code, to make sure they work. We really need them. 
You can see a bunch of them in moose/test/src

**Reply by aarograh** [reply]
For the first point, you're right, we have to do that for now.  It's just annoying to do it in two places (moleApp and moleTestApp).  It would make more sense to me if we could have moleTestApp inherit from moleApp, but for some reason it cannot identify the samApp objects when we do that.  I'd like to understand why this isn't working.

I think switching to dynamic loading might be the better approach long term, but we haven't worked that out yet.  Are their any guides/examples that we could look at to try this?

And we can check out the test objects in moose to see some examples.  Thanks for pointing us in that direction.

**Reply by GiudGiud** [reply]
If it fails to identify the samApp objects, then something went wrong during registration.
When you tried to inherit from MoleApp with MoleTestApp, what did you MoleTestApp registerApps and registerAll look like?

There arent many applications doing this because it is brittle when the two apps have been build with different MOOSE versions. There's not much documentation about it, I would just look at the tests for it in the misc module. 
You do it like this:
```
[Problem]
  register_objects_from = 'PhaseFieldApp'
  library_path = '../../../../../phase_field/lib'
[]
```
